### PR TITLE
WT-11957 Handle prepare conflicts when traversing next or prev with a bounded cursor

### PR DESCRIPTION
**Pull request by luke-pearson**
… bounded cursor

### CODE REVIEW

File: test/suite/test_cursor_bound21.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+import wiredtiger, wttest
+from wiredtiger import wiredtiger_strerror, WT_PREPARE_CONFLICT, WiredTigerError
+from wtscenario import make_scenarios
+from wtbound import bound_base
+
+# Test prepare conflict correctness when positioning a cursor for next and prev.
+#
+# test_cursor_bound21.py
+class test_cursor_bound21(bound_base):
+    conn_config = 'cache_size=2MB,eviction=(threads_max=1)'
```
Comment (jiechenbo) [inline test/suite/test_cursor_bound21.py:38]: do we need a really small cache size here?

**Review by jiechenbo — APPROVED** [review]
LGTM! Just some small changes.

File: src/btree/bt_curprev.c [hunk 2]
```diff
@@ -767,7 +767,23 @@ __wt_btcur_prev(WT_CURSOR_BTREE *cbt, bool truncating)
...
+         * If we get a prepare conflict we need to setup the correct retry flags in the case of
+         * row-store. Column-store doesn't require this as it always uses insert lists. If we don't
+         * set the appropriate flags we'll walk past the prepared update and violate prepared
+         * transaction semantics.
+         *
+         * We can use the cursor b-tree information to determine whether we are positioned on an
+         * insert list or a row. If an insert list is present then we ignore the on-page object.
+         */
+        if (ret == WT_PREPARE_CONFLICT && CUR2BT(cursor)->type == BTREE_ROW) {
+            if (cbt->ins != NULL)
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/btree/bt_curprev.c:786]: 7d682f081

File: test/suite/test_cursor_bound21.py (hunk 1 above)
Comment (luke-pearson) [reply test/suite/test_cursor_bound21.py]: 7d682f081

**Comment by tod-johnson-mongodb** [comment]
I have some questions and comments about some unchanged lines. github won't let me comment on unchanged lines, so I do it here.

- Question: Old 903-905, new 924-926: How does this if (... (... || (newpage && cbt->page_deleted_count > 0))) { match the comment "or we went all the way through a page and only saw deleted records"? Does cbt->page_deleted_cnt > 0 mean this page was deleted plus newpage mean time for next page thus implying this page had only delete records? Why does cbt->page_deleted_count > 0 refer to this page not a previous page?
- Possible bug: Old 918, New 939 "If ... and not interested in returning tombstones" does not match the code Old 924, New 945 if (... !F_ISSET(&cbt->iface, WT_CURSTD_IGNORE_TOMBSTONE)). Does not ignoring tombstones mean returning tombstones? If so, this is the opposite of the comment which says tombstones are not returned. Thus it seems to me the code should be if (... F_ISSET(&cbt->iface, WT_CURSTD_IGNORE_TOMBSTONE)).
-

**Comment by luke-pearson** [comment]
> Question: Old 903-905, new 924-926: How does this if (... (... || (newpage && cbt->page_deleted_count > 0))) { match the comment "or we went all the way through a page and only saw deleted records"? Does cbt->page_deleted_cnt > 0 mean this page was deleted plus newpage mean time for next page thus implying this page had only delete records? Why does cbt->page_deleted_count > 0 refer to this page not a previous page?

From what I can tell page_deleted_count is refreshed for every page visited in the walk, so as soon as we find a new page it is set to 0. I'm not entirely clear on the logic around `(newpage && cbt->page_deleted_count > 0)` @quchenhao would you be able to assist here?

> Possible bug: Old 918, New 939 "If ... and not interested in returning tombstones" does not match the code Old 924, New 945 if (... !F_ISSET(&cbt->iface, WT_CURSTD_IGNORE_TOMBSTONE)). Does not ignoring tombstones mean returning tombstones? If so, this is the opposite of the comment which says tombstones are not returned. Thus it seems to me the code should be if (... F_ISSET(&cbt->iface, WT_CURSTD_IGNORE_TOMBSTONE)).

The CURSTD_IGNORE_TOMBSTONE flag has this issue frequently, if set the flag indicates that the caller wants to ignore tombstones and read values even if they wouldn't have been able to read them. Because had they read the tombstone then the value would appear to be deleted. The comments says "not interested in returning tombstones" from what I can tell that is poorly written, we're not returning the tombstone so much as we choosing to ignore them. The `!` seems correct to me given that the comment also talks about snapshot isolation and ignoring tombstones breaks snapshot isolation.

**Comment by quchenhao** [comment]
> (newpage && cbt->page_deleted_count > 0)

I touched the code because I just reverted a change @hbokhari did (WT-10807). I also have no idea. It seems to be a bug to me.

This is the code from bt_curprev.c, which doesn't have that check. @hbokhari Can you explain why the check is only in the next code not in the prev code?
```
    /*
     * If we saw a lot of deleted records on this page, or we went all the way through a page and
     * only saw deleted records, try to evict the page when we release it. Otherwise repeatedly
     * deleting from the beginning of a tree can have quadratic performance. Take care not to force
     * eviction of pages that are genuinely empty, in new trees.
     *
     * A visible stop timestamp could have been treated as a tombstone and accounted in the deleted
     * count. Such a page might not have any new updates and be clean, but could benefit from
     * reconciliation getting rid of the obsolete content. Hence mark the page dirty to force it
     * through reconciliation.
     */
    if (cbt->page_deleted_count > WT_BTREE_DELETE_THRESHOLD) {
        WT_ERR(__wt_page_dirty_and_evict_soon(session, cbt->ref));
        WT_STAT_CONN_INCR(session, cache_eviction_force_delete);
    }
    cbt->page_deleted_count = 0;
```

**Comment by tod-johnson-mongodb** [comment]
LGTM

**Review by tod-johnson-mongodb — APPROVED** [review]

**Review by jiechenbo — APPROVED** [review]
LGTM!

**Comment by luke-pearson** [comment]
evergreen merge

**Comment by luke-pearson** [comment]
evergreen merge

**Pull request merged**